### ISSUE TITLE
Rearrange animation options logic

### ISSUE THREAD

**Issue by gibson042** [jquery/jquery#2255]
cc @gnarf

`.animate` uses the undocumented `jQuery.speed` (its only use, in fact) to manipulate options before sending them to `Animation`, where they are preserved as "original": https://github.com/jquery/jquery/pull/2218#issuecomment-98504382

We should move such manipulation so that `anim.originalOptions` is in fact what it claims to be.

**Comment by gnarf** [comment]
The problem is that method does a lot of the logic for the "argument options", I don't know how we could keep the "original options" in a `.slideToggle(200, complete)` case -- I really don't want to make `$.Animation` responsible for that.

*some* of the options manipulation (complete function) could probably be moved into the defaultPrefilter, but since `$.Animation` knows very little about the queue (I wish it knew less), I still think there needs to be a step in between `$.fn.animate` and it's aliases and the `$.Animation` options.

**Comment by gibson042** [comment]
I'm thinking that `$.fn.animate` would still bundle up speed/easing/callback into an `options` object for `$.Animate`, but that the object would be free of both default values and overrides—basically just the initial assignment from the current `$.speed`. Then the rest of the logic could be applied by `$.Animate`:
* converting duration to a number, zero if animations are off or instant
* defaulting to "fx" queue
* add a final step to `complete` for dequeueing

I can understand wanting to keep all of those external, though, especially as one moves down the list. What about adding an argument to it for such options preprocessing, to be used by `$.fn.animate`:
```js
// tail of jQuery.speed
function queueWrap( options ) {
	var opt = jQuery.extend( true, {}, options ),
		complete = opt.complete;

	// Default to "fx" queue
	if ( opt.queue == null || opt.queue === true ) {
		opt.queue = "fx";
	}

	// Dequeue when done
	if ( opt.queue ) {
		opt.complete = function() {
			if ( jQuery.isFunction( complete ) ) {
				complete.apply( this, arguments );
			}
			jQuery.dequeue( this, opt.queue );
		};
	}

	return opt;
}

// head of jQuery.speed
var options = speed && typeof speed === "object" ? jQuery.extend( {}, speed ) : {
	complete: callback || !callback && easing ||
		jQuery.isFunction( speed ) && speed,
	duration: speed,
	easing: callback && easing || easing && !jQuery.isFunction( easing ) && easing
};

// Animation initialization
var anim = Animation( this, jQuery.extend( {}, prop ), options, queueWrap )
```
?

There might even be opportunity to have the function take over some of what is now in `defaultPrefilter`.

**Comment by gnarf** [comment]
I like the idea, but I almost feel like passing `queueWrap( options )` makes more sense. Perhaps let `options.originalOptions` exist and be used by `$.Animation` ?

**Comment by Mr21** [comment]
Why not to put `$.speed` inside a seperate file to have this methode out of the public scope?
And maybe rename it by the way?

**Comment by gnarf** [comment]
I **think** we tried a few versions back to do that, but something somewhere used it.

I don't remember what, but I remember thinking it was a horrible function :smile:

**Comment by gnarf** [comment]
If we do anything, we should get rid of it...

**Comment by Mr21** [comment]
yeah, but it's not a probleme we can reuse the code in several files, no need to set this methode in public for that, no?

**Comment by gnarf** [comment]
We only use it one place (`jQuery.fn.animate`)... I think something else externally used it (jQuery UI maybe?)...  We shouldn't need this method at all if we do what @gibson042 is proposing.

**Comment by gnarf** [comment]
Doesn't take long to find people using it... https://github.com/search?l=javascript&q=%22jQuery+speed%22&ref=searchresults&type=Code&utf8=%E2%9C%93

* https://github.com/yckart/jquery-custom-animations/blob/b8ed1174b645ce4c02f6e1ca5c522579556f9302/js/jquery.custom-animations.js#L72
* https://github.com/rjollos/trachacks/blob/4fcd4aeba81d734654f5d9ec524218b91d54a0e1/tracforgeplugin/0.10/tracforge/htdocs/js/interface/ifxpulsate.js#L39

**Comment by gnarf** [comment]
Here's the jQuery UI: 
* https://github.com/jquery/jquery-ui/blob/6fd9fc92a372e222e3543f52979c601ac88ad81e/ui/effect.js#L785

**Comment by Mr21** [comment]
@gnarf 

> http://git.io/vTKOa

This one is not a good example because, he just add to remove the `jQuery.speed(` and it's works fine too. (exactly like here: http://git.io/vTKl3)

But, I am reading the jquery-ui's code that you send me.

**Comment by gibson042** [comment]
> Here's the jQuery UI: 
* https://github.com/jquery/jquery-ui/blob/6fd9fc92a372e222e3543f52979c601ac88ad81e/ui/effect.js#L785

Ugh. But on the bright side, https://github.com/jquery/jquery-ui/blob/6fd9fc92a372e222e3543f52979c601ac88ad81e/ui/effect.js#L1241 basically reproduces its non-queue aspects. _And_ it looks like UI's `animateClass` would only need [one change](https://github.com/jquery/jquery-ui/blob/6fd9fc92a372e222e3543f52979c601ac88ad81e/ui/effect.js#L854). I still think this can work.

> I like the idea, but I almost feel like passing `queueWrap( options )` makes more sense. Perhaps let `options.originalOptions` exist and be used by `$.Animation` ?

Possibly. It feels like a conflation of separate concerns, but I suppose it's no worse than passing the preprocessor. We'd just have to document, and thereby reserve, the property name—or just decide to nest rather than clone (e.g., `anim.originalOptions` would be the input to `$.Animate` and `anim.originalOptions.originalOptions` would be the input to `$.fn.animate`).

**Comment by gnarf** [comment]
`originalOptions` is only around for developer convenience, it's not used for anything we wrote.  I think it would be fine to just reuse the `originalOptions` passed into us if it is passed in.

**Comment by gibson042** [comment]
Yes, I agree, but wanted to make the decision explicit.

**Comment by Mr21** [comment]
The last time we add `easing._default`, but what would be the reason for a user to read the `originalOptions` instead of the `opts` attribute?
I didn't find an example on GitHub...

**Comment by gibson042** [comment]
`originalOptions` is an exact reference to user input, which might be valuable for two reasons:
* passing data outwards from animation callbacks
* retrieving configuration exactly as specified, before application of defaults/speed translations/etc.

We do the same thing with ajax calls.
